File: server/src/main/kotlin/com/coooolfan/unirhy/controller/SystemConfigController.kt
package com.coooolfan.unirhy.controller

import cn.dev33.satoken.annotation.SaCheckLogin
import com.coooolfan.unirhy.model.SystemConfig
import com.coooolfan.unirhy.model.by
import com.coooolfan.unirhy.model.dto.SystemConfigCreate
import com.coooolfan.unirhy.model.dto.SystemConfigUpdate
import com.coooolfan.unirhy.model.dto.SystemStatus
import com.coooolfan.unirhy.service.SystemConfigService
import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.sql.fetcher.Fetcher
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*

/**
 * 系统设置管理接口
 *
 * 提供系统配置的增删改查能力
 */
@RestController
@RequestMapping("/api/system/config")
class SystemConfigController(private val service: SystemConfigService) {

    /**
     * 获取系统初始化状态
     *
     * 此接口用于获取系统是否已完成初始化
     * 无需登录认证即可访问
     *
     * @return SystemStatus 返回系统初始化状态
     *
     * @api GET /api/system/config/status
     * @permission 不需要登录认证
     * @description 调用SystemConfigService.initialized()方法获取系统初始化状态
     */
    @GetMapping("/status")
    fun isInitialized(): SystemStatus {
        return SystemStatus(service.initialized())
    }

    /**
     * 获取系统配置
     *
     * 此接口用于获取当前系统配置
     * 需要用户登录认证才能访问
     *
     * @return SystemConfig 返回系统配置（默认 fetcher）
     *
     * @api GET /api/system/config
     * @permission 需要登录认证
     * @description 调用SystemConfigService.get()方法获取系统配置
     */
    @SaCheckLogin
    @GetMapping
    fun get(): @FetchBy("DEFAULT_SYSTEM_CONFIG_FETCHER") SystemConfig {
        return service.get(DEFAULT_SYSTEM_CONFIG_FETCHER)
    }

    /**
     * 创建系统配置
     *
     * 此接口用于创建系统配置（单例）
     * 需要用户登录认证才能访问
     *
     * @param create 创建参数
     * @return SystemConfig 返回创建后的系统配置（默认 fetcher）
     *
     * @api POST /api/system/config
     * @permission 需要登录认证
     * @description 调用SystemConfigService.create()方法创建系统配置
     */
    @SaCheckLogin
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun create(create: SystemConfigCreate): @FetchBy("DEFAULT_SYSTEM_CONFIG_FETCHER") SystemConfig {
        return service.create(create, DEFAULT_SYSTEM_CONFIG_FETCHER)
    }

    /**
     * 更新系统配置
     *
     * 此接口用于更新系统配置（单例）
     * 需要用户登录认证才能访问
     *
     * @param update 更新参数
     * @return SystemConfig 返回更新后的系统配置（默认 fetcher）
     *
     * @api PUT /api/system/config
     * @permission 需要登录认证
     * @description 调用SystemConfigService.update()方法更新系统配置
     */
    @SaCheckLogin
    @PutMapping
    fun update(update: SystemConfigUpdate): @FetchBy("DEFAULT_SYSTEM_CONFIG_FETCHER") SystemConfig {
        return service.update(update, DEFAULT_SYSTEM_CONFIG_FETCHER)
    }

    companion object {
        private val DEFAULT_SYSTEM_CONFIG_FETCHER: Fetcher<SystemConfig> =
            newFetcher(SystemConfig::class).by {
                allScalarFields()
                ossProviderId()
                fsProviderId()
            }
    }
}

File: web/src/__generated/services/SystemConfigController.ts
import type {Executor} from '../';
import type {SystemConfigDto} from '../model/dto/';
import type {SystemConfigUpdate, SystemInitReq, SystemStatus} from '../model/static/';

/**
 * 系统设置管理接口
 * 
 * 提供系统配置的增删改查能力
 */
export class SystemConfigController {
    
    constructor(private executor: Executor) {}
    
    /**
     * 初始化系统
     * 
     * 此接口用于初始化系统
     * 
     * @parameter {SystemConfigControllerOptions['create']} options
     * - create 创建参数
     * 
     */
    readonly create: (options: SystemConfigControllerOptions['create']) => Promise<
        void
    > = async(options) => {
        let _uri = '/api/system/config';
        return (await this.executor({uri: _uri, method: 'POST', body: options.body})) as Promise<void>;
    }
    
    /**
     * 获取系统配置
     * 
     * 此接口用于获取当前系统配置
     * 需要用户登录认证才能访问
     * 
     * @return SystemConfig 返回系统配置（默认 fetcher）
     * 
     */
    readonly get: () => Promise<
        SystemConfigDto['SystemConfigController/DEFAULT_SYSTEM_CONFIG_FETCHER']
    > = async() => {
        let _uri = '/api/system/config';
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<SystemConfigDto['SystemConfigController/DEFAULT_SYSTEM_CONFIG_FETCHER']>;
    }
    
    /**
     * 获取系统初始化状态
     * 
     * 此接口用于获取系统是否已完成初始化
     * 无需登录认证即可访问
     * 
     * @return SystemStatus 返回系统初始化状态
     * 
     */
    readonly isInitialized: () => Promise<
        SystemStatus
    > = async() => {
        let _uri = '/api/system/config/status';
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<SystemStatus>;
    }
    
    /**
     * 更新系统配置
     * 
     * 此接口用于更新系统配置（单例）
     * 需要用户登录认证才能访问
     * 
     * @parameter {SystemConfigControllerOptions['update']} options
     * - update 更新参数
     * @return SystemConfig 返回更新后的系统配置（默认 fetcher）
     * 
     */
    readonly update: (options: SystemConfigControllerOptions['update']) => Promise<
        SystemConfigDto['SystemConfigController/DEFAULT_SYSTEM_CONFIG_FETCHER']
    > = async(options) => {
        let _uri = '/api/system/config';
        return (await this.executor({uri: _uri, method: 'PUT', body: options.body})) as Promise<SystemConfigDto['SystemConfigController/DEFAULT_SYSTEM_CONFIG_FETCHER']>;
    }
}

export type SystemConfigControllerOptions = {
    'isInitialized': {}, 
    'get': {}, 
    'create': {
        /**
         * 创建参数
         * 
         */
        readonly body: SystemInitReq
    }, 
    'update': {
        /**
         * 更新参数
         */
        readonly body: SystemConfigUpdate
    }
}
